SOT-98: Summary of chapter 12 of the book Proofs, Arguments, and Zero-Knowledge (#35)
sota-zk-lab/zkp-documents: articles/proofs-arguments-and-zero-knowledge/chapter_12.md

Reviewer: General question: What should we remember after reading this chapter?

---

> Title: "\Sigma-Protocols and Commitments from Hardness of Discrete Logarithm"

Reviewer: ![Screenshot 2024-03-07 at 08 51 43](https://github.com/sota-zk-lab/zkp-documents/assets/33181397/35108522-9a14-499d-9eb9-047ba27d6fe8)

Status: Done
Level: "3"
---

# Chapter 12: $\Sigma$-Protocols and Commitments from Hardness of Discrete Logarithm

## 12.1 Cryptographic Background

### 12.1.1 A Brief Introduction to Groups

This is the definition of [group](../../terms/group.md).

### 12.1.2 The Discrete Logarithm Problem and Background on Elliptic Curves

#### 12.1.2.1 Discrete Log Problem

You should read the definition of Discrete Logarithm [here](../../terms/discrete_logarithm.md)

> #### 12.1.2.2 Elliptic Curve Groups

Reviewer: ```suggestion

#### 12.1.2.2 Elliptic Curve Groups
```

You should take a look at Elliptic curve group [here](../../terms/elliptic_curve_group.md).

## 12.2 Schnorr’s $\Sigma$-Protocol For Knowledge of Discrete Logarithms

This section covers:

> - Establishing that the prover has knowledge of a discrete logarithm of some group element ([Section 12.2.2](#12.2.2)).

Reviewer: ```suggestion
- Establishing that the prover has knowledge of a discrete logarithm of some group elements ([Section 12.2.2](#12.2.2)).
```

> - Allowing the prover to cryptographically commit to group elements without revealing the committed group element to the verifier until

Reviewer: ```suggestion
- Allowing the prover to cryptographically commit to group elements without revealing them to the verifier until
```

committer sends the opening information $e = m$ and $z$ to the verifier, who uses the verification procedure of the $\Sigma$-protocol to
confirm that $(a, e,z)$ is an accepting transcript for public input $h$.

By instantiating Damgård’s construction with Schnorr’s $\Sigma$-protocol for the discrete logarithm relation, one recovers a well-known
commitment scheme due to Pedersen. In the traditional presentation, to commit to a message $m$, the committer picks a random exponent
$z$ in $\{0,...,|G| −1\}$ and the commitment is $g^m  \cdot  h^z$ . One thinks of $h^z$ as a random group element that operates as a
“blinding factor”: by multiplying $g^m$ by $h^z$ , one ensures that the commitment is a random group element, statistically independent
of $m$. Below is the traditional presentation of Pedersen commitments:

---

**Protocol 5** Standard presentation of Pedersen commitments in a cyclic group $G$ for which the Discrete Logarithm problem is
intractable.

---

$1$: Let $G$ be a (multiplicative) cyclic group of prime order. The key generation procedure publishes randomly chosen generators $g,h
\in G$, which serve as both the commitment key and verification key.  

> $2$: To commit to a number $m \in \{0,...,|G| −1\}$, committer picks a random $z \in \{0,...,|G| −1\}$ and sends $c  \leftarrow{}  g^m

Reviewer 2: \lbrace and \rbrace

---
Title: "Σ-Protocols and Commitments from Hardness of Discrete Logarithm"
Status: Done
Level: "3"
---

# Chapter 12: $Σ$-Protocols and Commitments from Hardness of Discrete Logarithm

## 12.1 Cryptographic Background

### 12.1.1 A Brief Introduction to Groups

This is the definition of [group](../../terms/group.md).

### 12.1.2 The Discrete Logarithm Problem and Background on Elliptic Curves

> #### 12.1.2.1 Discrete Log Problem

Reviewer: Already have a section for this in discrete_logarithm.md

Reviewer: Merge this with it

Author: Should I add the text that implies "discrete logarithm problem are not post-quantum secure" to discrete_logarithm.md?

For a specified group $G$ the discrete logarithm problem takes as input two group elements $g$ and $h$, and the goal is to output a
positive integer $i$ such that $g^i = h$ (if $G$ is of prime order then such an $i$ is guaranteed to exist). The discrete logarithm
problem is believed to be computationally intractable in certain groups $G$. An important caveat is that quantum computers can solve
the discrete logarithm problem in polynomial time via Shor’s algorithm. Hence, cryptosystems whose security is based on the assumed
hardness of the discrete logarithm problem are not post-quantum secure.

> #### 12.1.2.2 Elliptic Curve Groups

Reviewer: Move this to `elliptic_curve_group.md` and link it with this chapter and the elliptic_curve definition

> You should take a look at Elliptic curve [here](../../terms/elliptic_curve.md). \

Reviewer: Missing definition of base field

---

> Title: "Σ-Protocols and Commitments from\r Hardness of Discrete Logarithm"

Reviewer: `\r`?

> Status: In Progress

Reviewer: In Progress ?

most elliptic curve groups used in practice runs in time $O(\sqrt{G})$. For example, a popular elliptic curve called Curve25519, which
is defined over base field $F$ of size $2^{255} −19$, defines a cyclic group of order close to $2^{252}$, so the time complexity is
about $O(2^{128})$.  
**Scalar field vs base field**: The field of size equal to the (prime) order of the elliptic curve group $G$ is typically referred to
as the _scalar field_ of $G$. It is not the same as the base field $F$ over which the curve is defined.

## 12.2 Schnorr’s $Σ$-Protocol For Knowledge of Discrete Logarithms

This section covers:

- Establishing that the prover has knowledge of a discrete logarithm of some group element
  ([Section 12.2.2](#12.2.2)).
- Allowing the prover to cryptographically commit to group elements without revealing the committed group element to the verifier until
  later ([Section 12.3](#12.3)).
- Establishing product relationships between committed values
  ([Section 12.3.2](#12.3.2))

### 12.2.1 $Σ$-Protocols

A relation $R$ specifies a collection of “valid” instance-witness pairs $(h,w)$. For example, $h=g^w$ for discrete logarithm relation.\

> A $\Sigma$-protocol is a 3-message public coin protocol in which bot verifier and prover know a public input $h$, and prover knows a

Reviewer: both

value $e'$ . Then run $P_{FS}$ once again to completion (again generating a random value from $R$’s response to each query made by
$P_{FS}$ ), and hope that $P_{FS}$ outputs an accepting transcript of the from $(a, e' ,z' )$.

**Analysis of the witness extraction procedure**: The probability this procedure outputs two accepting transcripts of the form $(a,
e,z)$ and $(a, e' ,z' )$ with $e \neq e'$ is at least $Ω(\varepsilon ^3/T^2 )$. Note that $e$ will not equal $e'$ with probability
$1−1/2^λ$ , where $λ$ denotes the number of bits output by $R$ on any query.

## 12.3 A Homomorphic Commitment Scheme<a id="12.3"></a>

You should read the definition of commitment scheme [here](../../terms/commitment_scheme.md).

**A Perfectly Hiding Commitment Scheme from any $Σ$-Protocol**: Damgård showed how to use any $Σ$-protocol for any hard relation to
obtain a perfectly hiding, computationally binding commitment scheme. To ensure hiding, Damgård’s transformation does require the
simulator used to establish HVZK must be able to take as input not only the public input $h$, but also a challenge $e^*$ , and output a

> transcript $(a, e^* ,z)$.

Reviewer 3: math problem

$3$: Prover picks $d,r ∈ \{0,...,|G| −1\}$ and sends to verifier $a ← Com(d,r).$  
$4$: Verifier sends challenge $e$ chosen at random from ${0,...,|G| −1}.$  
$5$: Let $m' ← me + d$ and $z' ← ze + r$, and let $c' ← Com(m ' ,z ' ).$ While Verifier does not know $m '$ and $z '$ , Verifier can
derive $c '$ unaided from $Com(m,z)$ and $Com(d,r)$ using additive homomorphism: $c'=Com(m,z)^e \cdot Com(d,r)$  
$6$: Prover sends $(m ' ,z' ).$  
$7$: Verifier checks that $m' ,z'$ is valid opening information for $c'$ , i.e., that $g^{m'} · h^{z'} = c'$ .

---

**Perfect Completeness**: $g^{m'} · h^{z'} = g^{me+d} · h^{ze+r} = c^e · a$ **Special Soundness**: Given two accepting transcripts $(a,
e,(m'_1 ,z'_1 ))$ and $(a, e',(m'_2 ,z'_2 ))$ with $e \neq e'$ , we have to extract a valid opening $(m,z)$ for the commitment $c$,
i.e., $g^m · h^z = c$. As in the analysis of the $Σ$-protocol for the Discrete Logarithm relation, we have:

$$
\begin{aligned} m^∗ = (m'_1 −m'_2 )·(e−e')^{−1} \quad mod \quad |G|, \\
z^∗ = (z'_1 −z'_2 )·(e−e')^{−1} \quad mod \quad |G|.
\end{aligned}
$$

Then:

$$

> \begin{aligned} g^{m^{*}}\cdot

Reviewer 3: math problem

---

$1$: Let $G$ be a (multiplicative) cyclic group of prime order over which the Discrete Logarithm relation is hard.  
$2$: Input is $c_i = g^{m_i} · h ^{r_i} = Com_{g,h}(m_i ,r_i)$ for $i ∈ \{1,2,3\}$ such that $m_3 = m_1 ·m_2$ mod $|G|.$  
$3$: Prover knows $m_i$ and $r_i$ for all $i ∈ \{1,2,3\},$ Verifier only knows $c_1, c_2, c_3,g,h$.  
$4$: Prover picks $b_1,...,b_5 ∈ \{0,...,|G| −1\}$ and sends to
verifier three values: $α ← Com_{g,h}(b_1,b_2), β ← Com_{g,h}(b_3,b_4), γ ← Com_{c_1,h}(b_3,b_5).$  
$5$: Verifier sends challenge $e$ chosen at random from $\{0,...,|G| −1\}.$  
$6$: Let $z_1 ← b_1 +e ·m_1, z_2 ← b_2 +e ·r_1, z_3 ← b_3 +e ·m_2, z_4 ← b_4 +e ·r_2, z_5 ← b_5 +e ·(r_3 −r_1m_2).$  
$7$: While Verifier does not know $z_1,...,z_5$, using additive homomorphism Verifier can derive the following three commitments
unaided using additive homomorphism:

$$
\begin{aligned} c_1' = Com_{g,h}(z_1,z_2) = α · c_1^e , \\
c_2' = Com_{g,h}(z_3,z_4) = β · c_2^e , \\
c_3' = Com_{c_1,h}(z_3,z_5) = γ · c_3^e \end{aligned}
$$

This final equality for $c_3'$ exploits that:

> $$ \begin{aligned} c*3^e = g^{em_1m_2} h^{er3} = c_1^{em_2} h^{er_3−er_1m_2} = Com*{c_1,h}(em_2, er_3 −er_1m_2). \end{aligned} $$

Reviewer 3: math problem

---
Title: "Σ-Protocols and Commitments from Hardness of Discrete Logarithm"
Status: Done
Level: "3"
---

# Chapter 12: $\Sigma$-Protocols and Commitments from Hardness of Discrete Logarithm

## 12.1 Cryptographic Background

### 12.1.1 A Brief Introduction to Groups

This is the definition of [group](../../terms/group.md).

### 12.1.2 The Discrete Logarithm Problem and Background on Elliptic Curves

#### 12.1.2.1 Discrete Log Problem

You should read the definition of Discrete Logarithm [here](../../terms/discrete_logarithm.md)

#### 12.1.2.2 Elliptic Curve Groups

You should take a look at Elliptic curve group [here](../../terms/elliptic_curve_group.md).

## 12.2 Schnorr’s $\Sigma$-Protocol For Knowledge of Discrete Logarithms

> This section covers:

Reviewer: This should be `the purpose of this sum protocol` or smth like that

Reviewer: ![Screenshot 2024-03-26 at 10 32 17](https://github.com/sota-zk-lab/zkp-documents/assets/33181397/412a0b34-a178-4c66-806c-c266bc84d11d)
![Screenshot 2024-03-26 at 10 32 27](https://github.com/sota-zk-lab/zkp-documents/assets/33181397/ff4a9acc-418a-4cf1-9b66-11aec8a2dcd6)

- Establishing that the prover has knowledge of a discrete logarithm of some group elements ([Section 12.2.2](#12.2.2)).
- Allowing the prover to cryptographically commit to group elements without revealing them to the verifier until
  later ([Section 12.3](#12.3)).
- Establishing product relationships between committed values ([Section 12.3.2](#12.3.2))

### 12.2.1 $\Sigma$-Protocols

A relation $R$ specifies a collection of “valid” instance-witness pairs $(h,w)$. For example, $h=g^w$ for discrete logarithm relation.\
A $\Sigma$-protocol is a 3-message public coin protocol in which both verifier and prover know a public input $h$, and prover knows a
witness $w$ such that $(h,w)\in R$. Let denote the three messages by $(a, e, z)$, with the prover first sending $a$, the verifier
responding with a challenge $e$ , and the prover replying with $z$. A $\Sigma$-protocol is required to satisfy perfect completeness,
and two
additional properties:

- **Special soundness**: There exists a polynomial time algorithm $Q$ such that, when given as input a pair of accepting transcripts
  $(a, e, z)$ and $(a, e',z')$ with $e \neq e'$ , $Q$ outputs a witness $w$ such that $(h,w) \in R$.

> - **Honest Verifier Perfect Zero-Knowledge (HVZK)**: There must be a randomized polynomial time simulator that takes as input the

Reviewer: This is hard to understand

  public input $h$ from the $\Sigma$-protocol, and outputs a transcript $(a, e,z)$ such that the distribution over transcripts output
  by the
  simulator is identical to the distribution over transcripts produced by the honest verifier in the $\Sigma$-protocol interacting with

>   the

Reviewer: Add **remark 12.1** section. It's important

  honest prover.

> ### 12.2.2 Schnorr’s $\Sigma$-Protocol For the Discrete Logarithm Relation <a id="12.2.2"></a>

Reviewer: <a id="12.2.2"></a> ?

The algorithm can be presented as follows:  

> $1$: Let $G$ be a (multiplicative) cyclic group of prime order with generator $g$.  

Reviewer: ```suggestion
1. Let $G$ be a (multiplicative) cyclic group of prime order with generator $g$.  
```

> $2$: Public input is $h$ = $g^w$, where only prover knows $w$.  

Reviewer: 2.

> $3$: Picks a random number $r \in \lbrace 0,...,|G| −1 \rbrace$ and sends $a \leftarrow{} g^r$ to the verifier.  

Reviewer: 3.

$4$: Verifier responds with a random element $e \in \lbrace 0,...,|G|−1 \rbrace.$  
$5$: Prover responds with $z \leftarrow{}  (we+r)$ mod $|G|$.  
$6$: Verifier checks that $a.h^e = g^z$.

> Perfect completeness is easy to establish, so we will prove that Schnorr’s protocol satisfies special soundness, and honest-verifier

Reviewer: ```suggestion
**Perfect completeness** is easy to establish, so we will prove that Schnorr’s protocol satisfies special soundness, and honest-verifier
```

> Perfect completeness is easy to establish, so we will prove that Schnorr’s protocol satisfies special soundness, and honest-verifier

Reviewer: Write down the formula

value $e'$ . Then run $P_{FS}$ once again to completion (again generating a random value from $R$’s response to each query made by
$P_{FS}$ ), and hope that $P_{FS}$ outputs an accepting transcript of the from $(a, e' ,z' )$.

**Analysis of the witness extraction procedure**: The probability this procedure outputs two accepting transcripts of the form $(a,
e,z)$ and $(a, e' ,z' )$ with $e \neq e'$ is at least $Ω(\varepsilon ^3/T^2 )$. Note that $e$ will not equal $e'$ with probability
$1−1/2^λ$ , where $λ$ denotes the number of bits output by $R$ on any query.

## 12.3 A Homomorphic Commitment Scheme<a id="12.3"></a>

You should read the definition of commitment scheme [here](../../terms/commitment_scheme.md).

**A Perfectly Hiding Commitment Scheme from any $\Sigma$-Protocol**: Damgård showed how to use any $\Sigma$-protocol for any hard
relation to
obtain a perfectly hiding, computationally binding commitment scheme. To ensure hiding, Damgård’s transformation does require the
simulator used to establish HVZK must be able to take as input not only the public input $h$, but also a challenge $e^{ * }$ , and
output a transcript $(a, e^ * ,z)$.

> Here is how Damgård’s commitment scheme works. Let denote $(h,w)  \leftarrow{} Gen$, and declares $h$ to be both the committing key

Reviewer: Split this into a subsection:
### Damgård’s commitment scheme

sends $a$ as the commitment, and keeps $e = m$ and $z$ as opening information. In the verification stage for the commitment scheme, the
committer sends the opening information $e = m$ and $z$ to the verifier, who uses the verification procedure of the $\Sigma$-protocol
to
confirm that $(a, e,z)$ is an accepting transcript for public input $h$.

By instantiating Damgård’s construction with Schnorr’s $\Sigma$-protocol for the discrete logarithm relation, one recovers a well-known
commitment scheme due to Pedersen. In the traditional presentation, to commit to a message $m$, the committer picks a random exponent
$z$ in $\lbrace 0,...,|G| −1 \rbrace$ and the commitment is $g^m \cdot h^z$ . One thinks of $h^z$ as a random group element that
operates as a
“blinding factor”: by multiplying $g^m$ by $h^z$ , one ensures that the commitment is a random group element, statistically independent
of $m$. Below is the traditional presentation of Pedersen commitments:

---

**Protocol 5** Standard presentation of Pedersen commitments in a cyclic group $G$ for which the Discrete Logarithm problem is
intractable.

---

> $1$: Let $G$ be a (multiplicative) cyclic group of prime order. The key generation procedure publishes randomly chosen generators $g,h

Reviewer: 1.

\in G$, which serve as both the commitment key and verification key.  
$2$: To commit to a number $m \in \lbrace 0,...,|G| −1 \rbrace$, committer picks a random $z \in \lbrace 0,...,|G| −1 \rbrace$ and
sends $c \leftarrow{} g^m
\cdot h^z$ .  
$3$: To open a commitment $c$, committer sends $(m,z)$. Verifier checks that $c = g^m \cdot h^z$ .

> ---

Reviewer: remove these `---`

> Pedersen commitment scheme obtained from Schnorr’s protocol via Damgård’s transformation:

Reviewer: Should be a header

> ---

Reviewer: ?

**Protocol 6** Commitment scheme obtained from Schnorr’s protocol via Damgård’s transformation.

---

> $1$: Let $G$ be a (multiplicative) cyclic group of prime order.  

Reviewer: 1.

$2$: The key generation procedure publishes randomly chosen generators $g,h \in G$, which serve as both the commitment key and
verification key.  
$3$: To commit to a number $m \in \lbrace 0,...,|G| −1 \rbrace$, committer picks a random $z \in \lbrace 0,...,|G| −1 \rbrace$ and
sends $c \leftarrow{}
h^{−m} \cdot g^z$  
$4$: To open a commitment $c$, committer sends $(m,z).$ Verifier checks that $c \cdot h^m = g^z$

---

### 12.3.1 Important Properties of Pedersen Commitments

**Additive Homomorphism**: This means that the verifier can take two commitments $c_1$ and $c_2$, to values $m_1,m_2 \in \lbrace
0,...,|G| −
1 \rbrace$ (with $m_1,m_2$ known to the committer but not to the verifier), and the verifier on its own can derive a commitment $c_3$
to $m_3
:= m_1 + m_2$, such that the prover is able to open $c_3$ to $m_3$. This is done by simply letting $c_3 \leftarrow{} c_1 \cdot
c_2$. As for "opening information" provided by the prover, if $c_1 = h^{m_1} \cdot g^{z_1}$ and $c_2 = h^{m_2} \cdot g^{z_2}$ ,
then $c_3 = h^{m_1+m_2} \cdot g^{z_1+z_2}$, so the opening information for $c_3$ is simply $(m_1 + m_2,z_1 + z_2).$

> **Perfect HVZK Proof of Knowledge of Opening**: It will be useful for the prover to prove that it knows how to open a commitment $c$ to

Reviewer: New sub section

some value, without actually opening the commitment. Let's see it:

---

**Protocol 7** Zero-Knowledge Proof of Knowledge of Opening of Pedersen Commitment

---

> $1$: Let $G$ be a (multiplicative) cyclic group of prime order over which the Discrete Logarithm relation is hard, with randomly chosen

Reviewer: Place algorithm in something like this:

```
abd
```

Author: Place these algorithms in block will raise math formula errors, I don't know a better way to deal with it.

a$.

---

---

**Protocol 8** Equivalent Exposition of Protocol 7 in terms of commitments and additive homomorphism.

---

$1$: Let $G$ be a (multiplicative) cyclic group of prime order over which the Discrete Logarithm relation is hard, with randomly chosen
generators $g$ and $h$.  
$2$: Let $Com(m,z)$ denote the Pedersen commitment $g^m \cdot h^z$ . Prover knows $m$ and $z$, Verifier only knows $Com(m,z),g,h$.  
$3$: Prover picks $d,r \in \lbrace 0,...,|G| −1 \rbrace$ and sends to verifier $a \leftarrow{} Com(d,r).$  
$4$: Verifier sends challenge $e$ chosen at random from ${0,...,|G| −1}.$  
$5$: Let $m' \leftarrow{} me + d$ and $z' \leftarrow{} ze + r$, and let $c' \leftarrow{} Com(m ' ,z ' ).$ While Verifier does not
know $m '$ and $z '$ , Verifier can derive $c '$ unaided from $Com(m,z)$ and $Com(d,r)$ using additive homomorphism: $c'=Com(m,z)^e
\cdot Com(d,r)$  
$6$: Prover sends $(m ' ,z' ).$  
$7$: Verifier checks that $m' ,z'$ is valid opening information for $c'$ , i.e., that $g^{m'} \cdot h^{z'} = c'$ .

---

**Perfect Completeness**: $g^{m'} \cdot h^{z'} = g^{me+d} \cdot h^{ze+r} = c^e \cdot a$

> **Special Soundness**: Given two accepting transcripts $(a, e,(m'_1 ,z'_1 ))$ and $(a, e',(m'_2 ,z'_2 ))$ with $e \neq e'$ , we have to

Reviewer: separate line

extract a valid opening $(m,z)$ for the commitment $c$, i.e., $g^m \cdot h^z = c$. As in the analysis of the $\Sigma$-protocol for the
Discrete Logarithm relation, we have:

$$ \begin{aligned} m^∗ = (m'_1 −m'_2 ) \cdot (e−e')^{−1} \quad mod \quad |G|, \\ z^∗ = (z'_1 −z'_2 ) \cdot (e−e')^{−1} \quad mod \quad
|G|.
\end{aligned} $$

Then:

$$ \begin{aligned} g^{m^{*}}\cdot h^{z^ *
}=\Big(g^{(m_{1}^{\prime}-m_{2}^{\prime})}h^{(z_{1}^{\prime}-z_{2}^{\prime})}\Big)^{(e-e^{\prime})^{-1}}=\Big(c^{e}\cdot
a\cdot\Big(c^{e^{\prime}}\cdot a\Big)^{-1}\Big)^{(e-e^{\prime})^{-1}}=c \end{aligned} $$

where the penultimate equality follows from the fact that $(a, e,(m'_1 ,z'_1 ))$ and $(a, e',(m'_1 ,z'_1 ))$ are accepting transcripts.
That is, $(m^* ,z^* )$ is a valid (message, opening information) pair for the commitment $c$.  
**Perfect HVZK**: The simulator samples $e,m' ,z'$ uniformly at random from $\lbrace 0,...,|G| −1 \rbrace$ and then sets $a
\leftarrow{} g^{m'}
\cdot h^{z'} \cdot
c^{−e}$ and outputs ($a, e,(m' ,z' )).$

> ### 12.3.2 Establishing A Product Relationship Between Committed Values<a id="12.3.2"></a>

Author: This section only writes about one content, so I think it should not be divided into some subsections
